src/eredis_client.erl

```diff
@@ -153,36 +153,51 @@ handle_info({ssl, Socket, Data}, #state{socket = Socket} = State) ->
     ok = ssl:setopts(Socket, [{active, once}]),
     {noreply, handle_response(Data, State)};
 
-handle_info({Type, Socket, _}, #state{socket = OurSocket} = State)
-  when OurSocket =/= Socket, Type =:= tcp;
-       OurSocket =/= Socket, Type =:= ssl ->
-    %% Ignore tcp messages when the socket in message doesn't match
-    %% our state. In order to test behavior around receiving
-    %% tcp_closed message with clients waiting in queue, we send a
-    %% fake tcp_close message. This allows us to ignore messages that
-    %% arrive after that while we are reconnecting.
-    {noreply, State};
-
-%% TCP socket errors
-handle_info({tcp_error, _Socket, _Reason}, State) ->
-    %% This will be followed by a close
-    {noreply, State};
-
-%% SSL socket errors
+%% Socket errors. If the network or peer is down, the error is not
+%% always followed by a tcp_closed.
+%%
 %% TLS 1.3: Called after a connect when the client certificate has expired
-handle_info({ssl_error, _Socket, Reason}, State) ->
+handle_info({Error, Socket, Reason},
+            #state{socket = Socket, transport = Transport} = State)
+  when Error =:= tcp_error; Error =:= ssl_error ->
+    Transport:close(Socket),
     maybe_reconnect(Reason, State);
 
 %% Socket got closed, for example by Redis terminating idle
 %% clients. If desired, spawn of a new process which will try to reconnect and
 %% notify us when Redis is ready. In the meantime, we can respond with
 %% an error message to all our clients.
-handle_info({tcp_closed, _Socket}, State) ->
-    maybe_reconnect(tcp_closed, State);
+%%
+%% fake_socket is used for testing.
+handle_info({Closed, Socket}, #state{socket = OurSocket} = State)
+  when Closed =:= tcp_closed orelse Closed =:= ssl_closed,
+       Socket =:= OurSocket orelse Socket =:= fake_socket ->
```

Comment:
nice, really good to visualize fake_socket here...it was a bit hidden

Reply:
should we change the state to match an undefined socket when closing?, i.e set #state{socket = undefined}

Reply:
ah, maybe_reconnect does that..fine